hub/core/dataset.py

```diff
@@ -485,3 +526,104 @@ def token(self):
         if self._token is None and self.path.startswith("hub://"):
             self._token = self.client.get_token()
         return self._token
+
+    @property
+    def _ungrouped_tensors(self) -> Dict[str, Tensor]:
+        """Top level tensors in this group that do not belong to any sub groups"""
+        return {
+            posixpath.basename(k): v
+            for k, v in self._tensors.items()
+            if posixpath.dirname(k) == self.group_index
+        }
+
+    @property
+    def _all_tensors_filtered(self) -> List[str]:
+        """Names of all tensors belonging to this group, including those within sub groups"""
+        return [
+            posixpath.relpath(t, self.group_index)
+            for t in self._tensors
+            if not self.group_index or t.startswith(self.group_index + "/")
+        ]
+
+    @property
+    def tensors(self) -> Dict[str, Tensor]:
+        """All tensors belonging to this group, including those within sub groups"""
+        return {t: self[t] for t in self._all_tensors_filtered}
+
+    @property
+    def _groups(self) -> List[str]:
+        """Names of all groups in the root dataset"""
+        meta_key = get_dataset_meta_key()
+        return self.storage.get_cachable(meta_key, DatasetMeta).groups  # type: ignore
+
+    @property
+    def _groups_filtered(self) -> List[str]:
+        """Names of all sub groups in this group"""
+        groups_filtered = []
+        for g in self._groups:
+            dirname, basename = posixpath.split(g)
+            if dirname == self.group_index:
+                groups_filtered.append(basename)
+        return groups_filtered
+
+    @property
+    def groups(self) -> Dict[str, "Dataset"]:
+        """All sub groups in this group"""
+        return {g: self[g] for g in self._groups_filtered}
+
+    def _is_root(self) -> bool:
+        return not self.group_index
+
+    @property
+    def parent(self):
+        if self._is_root():
+            return None
+        return Dataset(
+            self.storage,
+            self.index,
+            posixpath.dirname(self.group_index),
+            self.read_only,
+            self.public,
+            self._token,
+            self.verbose,
+        )
+
+    @property
+    def root(self):
+        if self._is_root():
+            return self
+        return Dataset(
+            self.storage,
+            self.index,
+            "",
+            self.read_only,
+            self.public,
+            self._token,
+            self.verbose,
+        )
+
+    def _create_group(self, name: str) -> "Dataset":
```

Comment:
what are create_group and _create_group doing differently? docstring?

Reply:
docstr added